File: scripts/taku_publisher/bundle.py
from __future__ import annotations

import json
import zipfile
from datetime import datetime, timezone
from hashlib import sha256
from pathlib import Path
from typing import Any

from .constants import SCHEMA_VERSION, SUPPORTED_RUNTIME_PLATFORMS
from .discovery import assert_publish_type_available
from .scanner import assert_scan_ready
from .util import PublisherError, atomic_write_bytes, atomic_write_json, read_json, save_state, sha256_bytes
from .workspace import assert_stage_unchanged


CAPABILITY_CONTRACT_VERSION = "0.2.0"
CAPABILITY_PACKAGE_SCHEMA_VERSION = "taku.package.v1"
CAPABILITY_SNAPSHOT_SCHEMA_VERSION = "taku.capability-snapshot.v1"
GENERATED_PACKAGE_MANIFEST_PATH = ".taku/package.json"
GENERATED_MANIFEST_PATH = ".taku/manifest.json"
GENERATED_REQUIREMENTS_PATH = ".taku/requirements.json"
KIND_ALIASES = {
    "skill": "skill",
    "skills": "skill",
    "agent": "agent",
    "agents": "agent",
    "subagent": "agent",
    "subagents": "agent",
    "action": "workflow",
    "actions": "workflow",
    "command": "workflow",
    "commands": "workflow",
    "slash-command": "workflow",
    "slash_command": "workflow",
    "workflow": "workflow",
    "workflows": "workflow",
    "plugin": "plugin",
    "plugins": "plugin",
}


def build_bundle(directory: Path, state: dict[str, Any], output_path: Path | None = None) -> dict[str, Any]:
    unit = state.get("unit") if isinstance(state.get("unit"), dict) else {}
    assert_publish_type_available(str(unit.get("type") or ""))
    manifest = assert_stage_unchanged(directory, state)
    assert_scan_ready(directory, state)
    staging = directory / "staging"
    requirements = read_json(directory / "requirements.json")
    bundle_manifest = _bundle_manifest(state)
    generated = {
        GENERATED_MANIFEST_PATH: _canonical_json(bundle_manifest),
        GENERATED_REQUIREMENTS_PATH: _canonical_json(requirements),
    }
    archive_path = output_path or directory / "bundle.zip"
    if archive_path.exists() and output_path is None:
        raise PublisherError("Bundle already exists for this draft.", code="bundle_exists")
    archive_bytes, bundle_files = _create_reproducible_zip(staging, manifest["files"], generated, state, requirements)
    bundle_sha256 = sha256_bytes(archive_bytes)
    atomic_write_bytes(archive_path, archive_bytes)
    if output_path is None:
        atomic_write_bytes(directory / "bundle.sha256", f"{bundle_sha256}  bundle.zip\n".encode("ascii"))
        file_list = dict(manifest)
        file_list["bundle_files"] = bundle_files
        file_list["bundle_file_count"] = len(bundle_files)
        atomic_write_json(directory / "file-list.json", file_list)
        state.update({
            "status": "packaged",
            "bundle_sha256": bundle_sha256,
            "bundle_size": len(archive_bytes),
            "bundle_file_count": len(bundle_files),
            "updated_at": datetime.now(timezone.utc).isoformat(),
        })
        save_state(directory, state)
    return {
        "path": str(archive_path),
        "sha256": bundle_sha256,
        "size": len(archive_bytes),
        "file_count": len(bundle_files),
        "files": bundle_files,
    }


def verify_local_bundle(directory: Path, state: dict[str, Any]) -> Path:
    assert_stage_unchanged(directory, state)
    bundle_path = directory / "bundle.zip"
    if not bundle_path.is_file():
        raise PublisherError("Bundle has not been created.", code="missing_bundle")
    actual = sha256_bytes(bundle_path.read_bytes())
    if actual != state.get("bundle_sha256"):
        raise PublisherError("Bundle digest no longer matches local state.", code="bundle_changed")
    return bundle_path


def _bundle_manifest(state: dict[str, Any]) -> dict[str, Any]:
    unit = state.get("unit") or {}
    children = unit.get("children") if isinstance(unit.get("children"), list) else []
    return {
        "schema_version": SCHEMA_VERSION,
        "type": unit.get("type"),
        "name": unit.get("name"),
        "description": unit.get("description") or "",
        "entrypoint": unit.get("entrypoint_relative") or Path(str(unit.get("entrypoint") or "")).name,
        "capabilities": [
            {
                "id": child.get("id"),
                "type": child.get("type"),
                "name": child.get("name"),
                "path": child.get("relative_path"),
            }
            for child in children
            if isinstance(child, dict)
        ],
    }


def _package_manifest(state: dict[str, Any], files: list[dict[str, Any]], requirements: dict[str, Any]) -> dict[str, Any]:
    unit = state.get("unit") if isinstance(state.get("unit"), dict) else {}
    normalized_files = sorted(files, key=lambda item: str(item.get("path") or "").casefold())
    raw_kind = str(unit.get("type") or "").strip()
    source_kind = _normalize_capability_kind(raw_kind)
    if not source_kind:
        raise PublisherError("Package manifest requires a supported capability kind.", code="invalid_package_manifest")
    if source_kind == "plugin":
        raise PublisherError(
            "Plugin packages require an approved taku.package.v1 permission review.",
            code="plugin_permission_review_required",
        )
    name = str(unit.get("name") or unit.get("id") or "").strip() or f"Untitled {source_kind}"
    description = str(unit.get("description") or "").strip()
    capability_id = str(unit.get("id") or "").strip() or _stable_capability_id(
        source_kind,
        "taku",
        str(unit.get("entrypoint_relative") or unit.get("entrypoint") or name),
    )
    required_secrets = sorted({
        str(item.get("name") or "").strip()
        for item in requirements.get("secrets", [])
        if isinstance(item, dict) and _is_secret_name(str(item.get("name") or "").strip())
    })
    content_hash = sha256(_canonical_capability_json({"files": normalized_files}).encode("utf-8")).hexdigest()
    capability: dict[str, Any] = {
        "id": capability_id,
        "kind": "action" if source_kind == "workflow" else source_kind,
        "sourceKind": source_kind,
        "name": name[:160],
    }
    if description:
        capability["description"] = description[:2000]
    return {
        "schemaVersion": CAPABILITY_PACKAGE_SCHEMA_VERSION,
        "contractVersion": CAPABILITY_CONTRACT_VERSION,
        "channel": "publish",
        "packageVersion": "1.0.0",
        "capability": capability,
        "contentHash": content_hash,
        "compatibility": {
            "hosts": sorted(SUPPORTED_RUNTIME_PLATFORMS),
            "platforms": sorted(SUPPORTED_RUNTIME_PLATFORMS),
        },
        "files": normalized_files,
        "permissions": [],
        "requiredSecrets": required_secrets,
    }


def _canonical_json(value: Any) -> bytes:
    return (json.dumps(value, ensure_ascii=False, indent=2, sort_keys=True) + "\n").encode("utf-8")


def _canonical_capability_json(value: Any) -> str:
    return json.dumps(_sort_deep(value), ensure_ascii=False, separators=(",", ":"))


def _sort_deep(value: Any) -> Any:
    if isinstance(value, list):
        return [_sort_deep(item) for item in value]
    if isinstance(value, dict):
        return {key: _sort_deep(value[key]) for key in sorted(value)}
    return value


def _normalize_capability_kind(value: str) -> str:
    return KIND_ALIASES.get(str(value or "").strip().lower(), "")


def _stable_capability_id(kind: str, source: str, locator: str) -> str:
    normalized_locator = str(locator or "").strip().replace("\\", "/").rstrip("/").lower()
    if not normalized_locator:
        normalized_locator = kind
    digest = sha256(f"{CAPABILITY_SNAPSHOT_SCHEMA_VERSION}\0{kind}\0{source}\0{normalized_locator}".encode("utf-8")).hexdigest()[:32]
    return f"cap_{digest}"


def _is_secret_name(value: str) -> bool:
    import re

    return bool(re.fullmatch(r"[A-Z][A-Z0-9_]{1,127}", value))


def _create_reproducible_zip(
    staging: Path,
    staged_files: list[dict[str, Any]],
    generated: dict[str, bytes],
    state: dict[str, Any],
    requirements: dict[str, Any],
) -> tuple[bytes, list[dict[str, Any]]]:
    import io

    memory = io.BytesIO()
    entries: list[tuple[str, bytes]] = []
    for entry in staged_files:
        relative = entry["path"]
        if relative in generated or relative.startswith(".taku/"):
            raise PublisherError("Source uses the reserved .taku package namespace.", code="reserved_package_path")
        entries.append((relative, (staging / relative).read_bytes()))
    entries.extend(generated.items())
    entries.sort(key=lambda item: item[0])
    package_files = [
        {
            "path": relative,
            "size": len(data),
            "sha256": sha256_bytes(data),
        }
        for relative, data in entries
    ]
    entries.append((
        GENERATED_PACKAGE_MANIFEST_PATH,
        _canonical_json(_package_manifest(state, package_files, requirements)),
    ))
    entries.sort(key=lambda item: item[0])

    file_list: list[dict[str, Any]] = []
    with zipfile.ZipFile(memory, mode="w", compression=zipfile.ZIP_STORED, strict_timestamps=True) as archive:
        for relative, data in entries:
            info = zipfile.ZipInfo(relative, date_time=(1980, 1, 1, 0, 0, 0))
            info.create_system = 3
            mode = 0o755 if _is_executable_script(relative) else 0o644
            info.external_attr = (mode & 0xFFFF) << 16
            info.compress_type = zipfile.ZIP_STORED
            archive.writestr(info, data)
            file_list.append({
                "path": relative,
                "size": len(data),
                "sha256": sha256_bytes(data),
            })
    return memory.getvalue(), file_list


def _is_executable_script(relative: str) -> bool:
    suffix = Path(relative).suffix.lower()
    return suffix in {".bash", ".command", ".sh", ".zsh"} or relative.startswith("bin/")

File: packages/publisher-runtime/src/bundle.ts
import * as fs from 'node:fs';
import * as fsp from 'node:fs/promises';
import * as path from 'node:path';

import {
  type CapabilityPackageFile,
  canonicalCapabilityJson,
  createCapabilityPackageManifest,
  stableCapabilityId,
} from '@taku/capability-contract';

import { SCHEMA_VERSION, SUPPORTED_RUNTIME_PLATFORMS } from './constants.js';
import { assertPublishTypeAvailable } from './discovery.js';
import { assertScanReady } from './scanner.js';
import type { FileManifestEntry, JsonObject, PublisherState } from './types.js';
import {
  atomicWriteBytes,
  atomicWriteJson,
  isRecord,
  nowIso,
  PublisherError,
  readJson,
  saveState,
  sha256Bytes,
} from './util.js';
import { assertStageUnchanged } from './workspace.js';
import { createStoredZip } from './zip.js';

const GENERATED_PACKAGE_MANIFEST_PATH = '.taku/package.json';
const GENERATED_MANIFEST_PATH = '.taku/manifest.json';
const GENERATED_REQUIREMENTS_PATH = '.taku/requirements.json';

export async function buildBundle(
  directory: string,
  state: PublisherState,
  outputPath?: string,
): Promise<JsonObject> {
  assertPublishTypeAvailable(String(state.unit.type ?? ''));
  const manifest = await assertStageUnchanged(directory, state);
  await assertScanReady(directory, state);
  const requirements = await readJson(path.join(directory, 'requirements.json'));
  if (!isRecord(requirements)) {
    throw new PublisherError('Requirements file is invalid.', 'invalid_requirements');
  }
  const generated = new Map<string, Buffer>([
    [GENERATED_MANIFEST_PATH, canonicalJson(bundleManifest(state))],
    [GENERATED_REQUIREMENTS_PATH, canonicalJson(requirements)],
  ]);
  const archivePath = outputPath ? path.resolve(outputPath) : path.join(directory, 'bundle.zip');
  if (!outputPath && fs.existsSync(archivePath)) throw new PublisherError('Bundle already exists for this draft.', 'bundle_exists');
  const files = Array.isArray(manifest.files) ? manifest.files.filter(isRecord) as FileManifestEntry[] : [];
  const { bytes, bundleFiles } = await createReproducibleZip(
    path.join(directory, 'staging'),
    files,
    generated,
    state,
    requirements,
  );
  const digest = sha256Bytes(bytes);
  await atomicWriteBytes(archivePath, bytes);
  if (!outputPath) {
    await atomicWriteBytes(path.join(directory, 'bundle.sha256'), Buffer.from(`${digest}  bundle.zip\n`, 'ascii'));
    const fileList: JsonObject = { ...manifest, bundle_files: bundleFiles, bundle_file_count: bundleFiles.length };
    await atomicWriteJson(path.join(directory, 'file-list.json'), fileList);
    state.status = 'packaged';
    state.bundle_sha256 = digest;
    state.bundle_size = bytes.length;
    state.bundle_file_count = bundleFiles.length;
    state.updated_at = nowIso();
    await saveState(directory, state);
  }
  return {
    path: archivePath,
    sha256: digest,
    size: bytes.length,
    file_count: bundleFiles.length,
    files: bundleFiles,
  };
}

function capabilityPackageManifest(
  state: PublisherState,
  files: CapabilityPackageFile[],
  requirements: JsonObject,
): JsonObject {
  const unit = isRecord(state.unit) ? state.unit : {};
  const kind = String(unit.type ?? '').trim();
  const name = String(unit.name ?? unit.id ?? '').trim();
  const description = String(unit.description ?? '').trim();
  const capabilityId = String(unit.id ?? '').trim() || stableCapabilityId({
    kind,
    source: 'taku',
    name,
    detectedFrom: String(unit.entrypoint_relative ?? unit.entrypoint ?? '').trim(),
  });
  const requiredSecrets = Array.isArray(requirements.secrets)
    ? requirements.secrets
      .filter(isRecord)
      .map((requirement) => String(requirement.name ?? '').trim())
      .filter(Boolean)
    : [];
  const contentHash = sha256Bytes(Buffer.from(
    canonicalCapabilityJson({
      files: [...files].sort((left, right) => left.path.localeCompare(right.path)),
    }),
    'utf8',
  ));
  return createCapabilityPackageManifest({
    channel: 'publish',
    packageVersion: '1.0.0',
    kind,
    contentHash,
    capability: {
      id: capabilityId,
      kind,
      name,
      description,
    },
    compatibility: {
      hosts: [...SUPPORTED_RUNTIME_PLATFORMS],
      platforms: [...SUPPORTED_RUNTIME_PLATFORMS],
    },
    files,
    permissions: [],
    requiredSecrets,
  }) as unknown as JsonObject;
}

export async function verifyLocalBundle(directory: string, state: PublisherState): Promise<string> {
  await assertStageUnchanged(directory, state);
  const bundlePath = path.join(directory, 'bundle.zip');
  if (!fs.existsSync(bundlePath)) throw new PublisherError('Bundle has not been created.', 'missing_bundle');
  const actual = sha256Bytes(await fsp.readFile(bundlePath));
  if (actual !== state.bundle_sha256) throw new PublisherError('Bundle digest no longer matches local state.', 'bundle_changed');
  return bundlePath;
}

function bundleManifest(state: PublisherState): JsonObject {
  const unit = isRecord(state.unit) ? state.unit : {};
  const children = Array.isArray(unit.children) ? unit.children.filter(isRecord) : [];
  return {
    schema_version: SCHEMA_VERSION,
    type: unit.type,
    name: unit.name,
    description: unit.description ?? '',
    entrypoint: unit.entrypoint_relative || path.basename(String(unit.entrypoint ?? '')),
    capabilities: children.map((child) => ({
      id: child.id,
      type: child.type,
      name: child.name,
      path: child.relative_path,
    })),
  };
}

function canonicalJson(value: unknown): Buffer {
  return Buffer.from(`${JSON.stringify(sortDeep(value), null, 2)}\n`, 'utf8');
}

async function createReproducibleZip(
  staging: string,
  stagedFiles: FileManifestEntry[],
  generated: Map<string, Buffer>,
  state: PublisherState,
  requirements: JsonObject,
): Promise<{ bytes: Buffer; bundleFiles: JsonObject[] }> {
  const entries: Array<{ name: string; data: Buffer; mode: number }> = [];
  for (const entry of stagedFiles) {
    if (generated.has(entry.path) || entry.path.startsWith('.taku/')) {
      throw new PublisherError('Source uses the reserved .taku package namespace.', 'reserved_package_path');
    }
    entries.push({
      name: entry.path,
      data: await fsp.readFile(path.join(staging, entry.path)),
      mode: executableScript(entry.path) ? 0o755 : 0o644,
    });
  }
  for (const [name, data] of generated) entries.push({ name, data, mode: 0o644 });
  entries.sort((left, right) => left.name < right.name ? -1 : left.name > right.name ? 1 : 0);
  const packageFiles = entries.map((entry) => ({
    path: entry.name,
    size: entry.data.length,
    sha256: sha256Bytes(entry.data),
  }));
  const packageManifest = capabilityPackageManifest(
    state,
    packageFiles,
    requirements,
  );
  entries.push({
    name: GENERATED_PACKAGE_MANIFEST_PATH,
    data: canonicalJson(packageManifest),
    mode: 0o644,
  });
  entries.sort((left, right) => left.name < right.name ? -1 : left.name > right.name ? 1 : 0);
  const bundleFiles = entries.map((entry) => ({
    path: entry.name,
    size: entry.data.length,
    sha256: sha256Bytes(entry.data),
  }));
  return { bytes: createStoredZip(entries), bundleFiles };
}

function executableScript(relative: string): boolean {
  return ['.bash', '.command', '.sh', '.zsh'].includes(path.extname(relative).toLowerCase())
    || relative.startsWith('bin/');
}

function sortDeep(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(sortDeep);
  if (!value || typeof value !== 'object') return value;
  return Object.fromEntries(
    Object.entries(value as Record<string, unknown>)
      .sort(([left], [right]) => left.localeCompare(right))
      .map(([key, child]) => [key, sortDeep(child)]),
  );
}
